File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>>
    {
        let mut last = None;
        let mut head = head;

        while let Some(mut node) = head
        {
            let next = node.next;
            node.next = last;
            (last,head) = (Some(node),next);
        }
        return last;
    }
}

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cpp

class Solution {
public:
    bool hasCycle(ListNode* head)
    {
        auto fast = head;
        auto slow = head;

        while (true)
        {
            if (fast == nullptr || fast->next == nullptr) {
                return false;
            }

            fast = fast->next->next;
            slow = slow->next;

            if (fast == slow) {
                return true;
            }
        }
    }
};
